### PR TITLE
Refactored validator to interface

### PR DESCRIPTION
**Pull request by jmatosp**
Added Validator interface with all methods from validator struct
Added return errors in methods to gracefully handle them
Logic found in handler when receiving messages (validator.Validating) is now in the Validator it self, and an error is raised if not validating (allows to handle the error), please verify this one as is a slight change in program flow (handler.go)

Important changes:
now most methods might return an identifiable error

### CODE REVIEW

File: kusd/handler.go [hunk 1]
```diff
@@ -628,48 +628,36 @@ func (pm *ProtocolManager) handleMsg(p *peer) error {
...
 		var proposal types.Proposal
 		if err := msg.Decode(&proposal); err != nil {
 			return errResp(ErrDecode, "msg %v: %v", msg, err)
 		}
-		pm.validator.AddProposal(&proposal)
-
-	case msg.Code == VoteMsg:
-		// @TODO (rgeraldes) - review flow (we will not need this condition)
-		if !pm.validator.Validating() {
+		if err := pm.validator.AddProposal(&proposal); err != nil {
```
Comment (acroca) [inline kusd/handler.go]: Should we add some log if this returns an `err`?

File: kusd/handler.go [hunk 2]
```diff
@@ -628,48 +628,36 @@ func (pm *ProtocolManager) handleMsg(p *peer) error {
...
 			return errResp(ErrDecode, "msg %v: %v", msg, err)
 		}
-
-		p.MarkVote(vote.Hash())
-		pm.validator.AddVote(&vote)
-
-	case msg.Code == BlockFragmentMsg:
-		// @TODO (rgeraldes) - review flow (we will not need this condition)
-		if !pm.validator.Validating() {
+		if err := pm.validator.AddVote(&vote); err != nil {
```
Comment (acroca) [inline kusd/handler.go]: Same here, should we log?

File: kusd/validator/validator.go [hunk 3]
```diff
@@ -31,8 +32,30 @@ type Backend interface {
...
+	SetDeposit(deposit uint64)
+	Pending() (*types.Block, *state.StateDB)
+	PendingBlock() *types.Block
+	AddProposal(proposal *types.Proposal) error
+	AddVote(vote *types.Vote) error
+	AddBlockFragment(blockNumber *big.Int, round uint64, fragment *types.BlockFragment) error
+}
+
+var (
+	ErrCantStopNonStartedValidator       = errors.New("can't stop consensus, not started")
```
Comment (acroca) [inline kusd/validator/validator.go]: just for consistency, shouldn't this be `not validating`?

File: kusd/handler.go [hunk 4]
```diff
@@ -628,48 +628,36 @@ func (pm *ProtocolManager) handleMsg(p *peer) error {
...
 
+	case msg.Code == BlockFragmentMsg:
 		// Retrieve and decode the propagated block fragment
 		var request blockFragmentData
 		if err := msg.Decode(&request); err != nil {
 			return errResp(ErrDecode, "msg %v: %v", msg, err)
 		}
-
-		// @TODO (rgerades) - fragment hash?
+		if err := pm.validator.AddBlockFragment(request.BlockNumber, request.Round, request.Data); err != nil {
```
Comment (acroca) [inline kusd/handler.go]: Same here, log the error?

File: kusd/handler.go (hunk 1 above)
Comment (jmatosp) [reply kusd/handler.go]: I think so, either log or comment saying why we ignoring the error

File: kusd/handler.go (hunk 2 above)
Comment (jmatosp) [reply kusd/handler.go]: :+1:

File: kusd/handler.go (hunk 4 above)
Comment (jmatosp) [reply kusd/handler.go]: :+1:

File: kusd/validator/validator.go (hunk 3 above)
Comment (jmatosp) [reply kusd/validator/validator.go]: :+1:  true one or the other, not sure which though

File: kusd/handler.go (hunk 1 above)
Comment (rgeraldes) [reply kusd/handler.go]: on the same file pm.txpool.AddRemotes(txs) throws an error but they are not handling the error. Have a look at their flow to see if we can do something similar.

File: kusd/validator/validator.go [hunk 5]
```diff
@@ -30,750 +18,16 @@ type Backend interface {
...
+	Stop() error
+	SetExtra(extra []byte) error
+	Validating() bool
+	SetCoinbase(addr common.Address) error
+	SetDeposit(deposit uint64)
+	Pending() (*types.Block, *state.StateDB)
+	PendingBlock() *types.Block
+	AddProposal(proposal *types.Proposal) error
+	AddVote(vote *types.Vote) error
+	AddBlockFragment(blockNumber *big.Int, round uint64, fragment *types.BlockFragment) error
```
Comment (rgeraldes) [inline kusd/validator/validator.go]: LIKE!

File: kusd/handler.go (hunk 1 above)
Comment (jmatosp) [reply kusd/handler.go]: not sure if I follow, they are ignoring the errors, what should we do then?

File: kusd/handler.go [hunk 6]
```diff
@@ -628,48 +628,39 @@ func (pm *ProtocolManager) handleMsg(p *peer) error {
 		pm.txpool.AddRemotes(txs)
 
 	case msg.Code == ProposalMsg:
-		// @TODO (rgeraldes) - review flow (we will not need this condition)
-		if !pm.validator.Validating() {
```
Comment (rgeraldes) [inline kusd/handler.go]: The flow changed a little bit - We don't need to decode the message if we're not validating. 
With the new flow we decode the message even if it's not necessary.

Comment (jmatosp) [reply kusd/handler.go]: this should be handled in a diff way, but I see why we need the Validating(), adding back

**Pull request merged**